single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+- deploy a new instance of `retagger` that will replicate all the configured public images to the new `gsoci` registry
+- switch container runtimes to use the cache as a source of images and upstream `gsoci` as a fallback
+- cleanup (after some time)
+  - delete our registries in docker hub and quay
+  - delete the old `giantswarm` registry in ACR
+- for this part, we keep China registry unchanged
+  - we reconfigure the `crsync` to replicate images from `gsoci` to `aliyun` in China
+
+#### China extension
+
+- we try to setup automated geo-replication between Europe and China
+- switch China clusters to use the new ACR-China registry
+- shut down `crsync` entirely
+- delete Aliyun registry
+
+### Helm charts
+
+The blocker for switching Helm Charts entirely to OCI is currently our artifact discovery process, which uses the
+`index.yaml` file, that is available in HTTP based Helm registries only. So, to get started, we have to switch our
+discovery process to OCI based one. Currently, we're already uploading our build artifacts to both HTTP and OCI repos,
+so nothing is blocking implementation and a switch.
+
+High level migration plan:
+
+- prepare and deploy `distribution` cache instance, configure it for the upstream ACR registry
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - we want to use retention policies for helm charts as well
+- implement changes in the `app-operator`
+  - use OCI discovery instead of `index.yaml` from HTTP registries
+  - (optional) solve the problem of presenting only X (by default 5) most recent versions in the ACE CR
+  - introduce a concept of primary and fall-back registry
+- stop uploading charts into HTTP registries
+- reconfigure `app-operator` to use the OCI cache as the primary registry and upstream ACR as fallback
+- cleanup HTTP registries
+- handle China as for images
+
+## Open questions
+
+Before we start on the implementation, we have to figure out answers to questions below:
+
+1. Should we use a single cache instance per installation, or rather per cloud provider + region?
+
+   The disk requirements for a cache might be significant and as a result cost of running an instance per MC can be quite
+   high in total. Additionally, we can expect many of the images to be the same for all clusters (MCs and WCs) that run
+   for the same provider. So, it seems running a single instance of the cache per cloud provider can be both much cheaper
+   and also more efficient and performant.
+
+1. How to not tag dev builds?
+
+   ACR has a neat new feature that allows for automatic cleanup of untagged artifacts after a certain period. Currently,
+   we upload a ton of dev build images. This feature allows us to clean them up automatically, with zero effort on
+   the cleaning process on our side. Still, we have to switch our CI/CD configs to not tag dev builds in any way. What's
+   the best way to do it?
```

Comment:
I guess we can reference with hash, but I am not sure app platform supports it, e.g. AppCR.spec.version.

Reply:
Can we think of another cleanup strategy that's easy to implement with the ACR built-in features? One where dev builds can have tags, too. I'm asking for the case that we find we want tags on dev artifacts.

Reply:
We can have a tagging schema that we can use to detect "dev" builds and then write a piece of software, that runs and "untags" dev builds. But that's again a service we have to write to maintain, and I want to use as few custom pieces we have to support as possible.

Reply:
If we need such a software, maybe we can utilise OCI here to store that something is a dev build.

Reply:
Do you mean some meta-data associated with an image? Sure, that's a valid idea.